### DISCUSSION TITLE
MAUI memory use and pages not get garbage collected?

### DISCUSSION THREAD

**Discussion by JoacimWall** [dotnet/maui#20125] [Q&A]
Hello
We have an app running on latest MAUI 8.0.6 same for 8.0.3 that just keeps growing and growing in memory when we run it in debug or Release. I know you should test this in release mode but can't get it to work.

Another thing is that XCode Instruments doesn't seem to work with MAUI.

Have also tried running the same app via the Apple App Store as macCatalyst in Release mode sees the same behaviour in the Activity Monitor on Mac where the memory usage only increases. I can see that if I run 
GC.WaitForPendingFinalizers();
GC.Collect();
and 
GC.GetGCMemoryInfo();
the memory decrease but I think the memory that increases is not handled by the dotnet GC but mono perhaps or something native?

So I created a simple test app:
This test app is only a main page and a second page with an image. 
I bind Image.Source to an image and navigate back and forth and the memory usage just goes up. 

If I decide to set Image.source = null in OnDisappearing then the memory usage does not grow as much but still it seems that the page is not handled by the GC. 

The app increases with 70mb for showing a image that is 3mb every time I navigate to the page.

In case we try to make a similar app in Xcode native swift. The same app use 16 mb instead of 240 mb for MAUI. And the app is not increasing in memory use in Activity monitor.  

So my questions is has anyone else seen this and is it something that I missing.

//Joacim

**Comment by yurkinh** [comment]
Hi @JoacimWall 
Yes, there are still some memory leaks in the basic controls.
Try this: https://github.com/AdamEssenmacher/MemoryToolkit.Maui
It should help to find you leaky control

**Reply by Xyncgas** [reply]
Why not use Visual Studio's profiling to find leaks
Meanwhile you can try MAUI.Blazor which is webUI that has a bunch web UI libraries everywhere without memory leaking

**Reply by Suundumused** [reply]
The above solution does not solve the original problem, I believe the problem is related to the topic: https://github.com/dotnet/maui/issues/17049

Which was closed early and treated as possibly solved with .NET 9, as if the .NET 8 version was out of support (by the way the problem was addressed there), the topic was closed without any modification or chance to comment / test.

**Comment by Suundumused** [comment]
I also have terrible problems with GC in MAUI, which is practically non-existent, every time you open and close a page, more and more is accumulated in memory, and little is unloaded, the load / unload ratio is always in deficit, even manually calling the GC periodically, it seems that the main problem is that the references are not abandoned.

**Reply by stephenquan** [reply]
You need to add further clarification to "every time" you open and close a page. If we're taking a handful number of times, you will definitely not notice a GC, nor will notice memory being unloaded. If we're talking 100s or even 1000s of times, then, definitely, I would expect some of your early page instances to have been either recycled or unloaded. 

There are things that actually can prevent a page from being unloaded. but, we would need to see the .NET code to identify which references aren't being released.

**Reply by Suundumused** [reply]
`Always` I mean `when`, not `rate` as it would be impossible for the MAUI interface to handle opening and closing pages at a rate greater than 10s with the most powerful hardware.

I understand that if you open a page and then close it, part of it is loaded to be used as cache and then it opens again faster, but when you close and open the same page several times and the memory usage adds up to the same amount previously, then there is a problem with `await Navigation.PushAsync(new template1());`, it should discard the previous reference.

I noticed that when upgrading from **.NET 8** to **9**, the debug log output started to notify GC flooding but without considerable discarding of objects.
